src/foam/core/CSV.js

```diff
@@ -0,0 +1,233 @@
+
+  documentation: 'CSV Outputter.',
+
+  properties: [
+    {
+      class: 'String',
+      name: 'buf_',
+      value: ''
+    },
+    {
+      class: 'String',
+      name: 'delimiter',
+      value: ','
+    },
+    {
+      class: 'String',
+      name: 'nestedObjectSeperator',
+      value: '__'
+    },
+    {
+      class: 'String',
+      name: 'nlStr',
+      value: '\n'
+    },
+    {
+      class: 'String',
+      name: 'undefinedStr',
+      value: ''
+    },
+    {
+      class: 'Boolean',
+      name: 'outputHeaderRow',
+      value: true
+    },
+    {
+      class: 'Function',
+      name: 'propertyPredicate',
+      value: function(o, p) { return ! p.transient; }
+    }
+  ],
+
+  methods: [
+    function out() {
+      for ( var i = 0 ; i < arguments.length ; i++ ) { 
+        this.buf_ += arguments[i];
+      }
+
+      return this;
+    },
+
+    function outputHeader(o) {
+      if ( ! this.outputHeaderRow ) return;
+
+      this.outputPropertyFilteredName(o, '');
+      this.out(this.nlStr);
+    },
+
+    function toCSV(o) {
+      // Outputs object headers
+      this.outputHeader(o);
+      var ret = this.buf_;
+      this.reset(); // reset to avoid retaining garbage
+
+      // Outputs object values
+      this.output(o);
+      this.out(this.nlStr);
+      ret += this.buf_;
+      this.reset();
+
+      return ret;
+    },
+
+    function outputHeaderTitle(o, prefix) {
+      this.out(this.buf_.length ? this.delimiter : '')
+          .out(prefix, o);
+    },
+
+    function outputPropertyName(o, p, prefix) {
+      if ( ! this.propertyPredicate(o, p) ) return;
+
+      // Checks if property is enum, or object with properties of its own
+      if ( p.of && ( ! foam.core.AbstractEnum.isInstance( o[p.name] ) ) ) {
+        // Appends object name to prefix for CSV Header
+        prefix += p.name + this.nestedObjectSeperator;
+        this.outputPropertyFilteredName(o[p.name], prefix);
+      } else {
+        this.outputPropertyFilteredName(p.name, prefix);
+      }
+    },
+
+    {
+      name: 'outputPropertyFilteredName',
+      code: foam.mmethod({
+        FObject:   function(o, prefix) {
+          // Gets and recurses through object properties
+          var ps = o.cls_.getAxiomsByClass(foam.core.Property);
+
+          for ( var i = 0 ; i < ps.length ; i++ ) {
+            this.outputPropertyName(o, ps[i], prefix);
+          }
+        },
+        Array: function(o) { /* Ignore arrays in CSV */ },
+        Function: function(o) { /* Ignore functions in CSV */ },
+        Object: function(o) { /* Ignore generic objects in CSV */ }
+      }, function(o, prefix) { this.outputHeaderTitle(o, prefix); })
+    },
+
+    function outputProperty(o, p) {
+      if ( ! this.propertyPredicate(o, p) ) return;
+      this.output(o[p.name]);
+    },
+
+    function reset() {
+      this.buf_ = '';
+      return this;
+    },
+
+    function outputPrimitive(val) {
+      this.out(this.buf_.length ? this.delimiter : '', val);
```

Comment:
I think it would be easier to handle the delimiter where you're doing the looping.